dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/DiagnosticsHelper.java

```diff
@@ -0,0 +1,387 @@
+
+        environment = newEnvironment;
+        sessionFactory = newSessionFactory;
+        config = newConfig;
+
+        if (config.getLogPeriodic()) {
+            // Create a daemon-thread-backed Timer and schedule a periodic dump of the global information.
+            long periodSeconds = config.getPeriodSeconds();
+            long periodMilliseconds = Math.round(periodSeconds * MILLISECONDS_PER_SECOND);
+            new Timer("diagnostics", true).scheduleAtFixedRate(
+                new TimerTask() {
+                    public void run() {
+                        logGlobals();
+                    }
+                }, periodMilliseconds, periodMilliseconds);
+            LOG.info(String.format("logging diagnostic information every %d seconds", periodSeconds));
+        }
+
+        if (config.getLogRequests()) {
+            // Register a Jersey event handler that will log session and thread-related information.
+            environment.jersey().register(new DiagnosticsHelperApplicationEventListener());
+            // Register a Jersey filter to determine the resource method.
+            environment.jersey().register(new DiagnosticsHelperContainerRequestFilter());
+            LOG.info("logging diagnostic request information");
+        }
+    }
+
+    public void logGlobals() {
+        logProcesses();
+        logFilesystems();
+        logDatabase();
+        logMemory();
+    }
+
+    public void logThreads() {
+        log("threads", () -> formatThreads());
+    }
+
+    public void logProcesses() {
+        log("processes", () -> formatProcesses());
+    }
+
+    public void logFilesystems() {
+        log("filesystems", () -> formatFilesystems());
+    }
+
+    public void logDatabase() {
+        log("database", () -> formatDatabase());
+    }
+
+    public void logMemory() {
+        log("memory", () -> formatMemory());
+    }
+
+    public void logStart(ContainerRequest request) {
+        log("start", () -> formatRequest(request));
+    }
+
+    public void logFinish(ContainerRequest request, ContainerResponse response, ThreadState startThreadState, ThreadState finishThreadState, Optional<SessionState> sessionState, Optional<User> user, Optional<Method> resourceMethod) {
+        log("finish", () -> formatRequest(request)
+            + formatUser(user)
+            + formatResourceMethod(resourceMethod)
+            + formatResponse(response)
+            + formatThread(startThreadState, finishThreadState)
+            + formatSession(sessionState));
+    }
+
+    public void log(String type, Supplier<String> valueSupplier) {
+        if (logger.isInfoEnabled()) {
+            Thread current = Thread.currentThread();
+            String message = String.format("diagnostics.%s by thread \"%s\" (%s):\n%s", type, current.getName(), current.getId(), valueSupplier.get());
+            logger.info(censorHelper.censor(message));
```

Comment:
> Diagnostic information should probably be `trace`

That's fine if we can configure the logging system in our deployments to include this output, but not the other TRACE/DEBUG output.  The diagnostic output is most useful if it's always logged, so we can refer back to it to debug problems, but we don't want to cloud the logs with TRACE/DEBUG output from other modules.

Reply:
> That's fine if we can configure the logging system in our deployments to include this output

This is pretty simple https://dev.to/cerveada/set-logback-to-different-level-for-specific-package-19db

> The diagnostic output is most useful if it's always logged 

This is configurable. If you have trace or debug on, these messages will always show up or vice versa. My intent here is to properly categorize the information 

i.e. `INFO` is normally pretty broad in scope `The INFO level designates informational messages that highlight the progress of the application at coarse-grained level.`
This sounds a lot more like `The TRACE Level designates finer-grained informational events than the DEBUG`
https://logging.apache.org/log4j/1.x/apidocs/org/apache/log4j/Level.html for example

Reply:
Actually, `logback-test.xml` may be just when running tests locally. 
[dropwizard config ](https://www.dropwizard.io/en/release-3.0.x/manual/configuration.html#logging) has the same idea for customizing logging by package

Reply:
> > That's fine if we can configure the logging system in our deployments to include this output
> 
> This is pretty simple https://dev.to/cerveada/set-logback-to-different-level-for-specific-package-19db
> 
> > The diagnostic output is most useful if it's always logged
> 
> This is configurable. If you have trace or debug on, these messages will always show up or vice versa. My intent here is to properly categorize the information
> 
> i.e. `INFO` is normally pretty broad in scope `The INFO level designates informational messages that highlight the progress of the application at coarse-grained level.` This sounds a lot more like `The TRACE Level designates finer-grained informational events than the DEBUG` https://logging.apache.org/log4j/1.x/apidocs/org/apache/log4j/Level.html for example

I think we differ philosophically here, I see `TRACE` and `DEBUG` as applying to tidbits of information that we're not interested in seeing most of the time, that we only enable to help us understand the behavior of some localized bit of code, whereas the information we're logging in this PR is relevant globally, will address some very significant blind spots, and help us to determine the cause of (and devise solutions to) a broad range of issues.  The additional information in this PR will help me _a lot_, with respect to both adding features and solving problems.

However, yes, we can change the log level to whatever the team wants, if we can make sure the display of the new information doesn't trigger the display of the less-important stuff...

Reply:
Did the output change a lot from what you had documented in https://github.com/dockstore/dockstore/pull/5821#issuecomment-1969954529 ?

I'm seeing timing information, thread information, memory usage, periodic `ps` output and the helper is/was literally called "io.dockstore.webservice.helpers.DebugHelper" i.e. `debug`? 

>  if we can make sure the display of the new information doesn't trigger the display of the less-important stuff...

I also don't think it helps to think about it in terms of importance. Debugging info can be really important depending on what you're working on, so can `info`. In other contexts, they can both be noise.

Again, I'm more thinking about categorisation.  The convention is normally that stuff like heap vs non-heap memory, connection pool size, etc. is more about behind-the-scenes detail as this article puts it

> A INFO log is like the movie trailer, giving you the overall plot. DEBUG is the director's commentary, pointing out significant scenes and decisions. TRACE, on the other hand, includes every behind-the-scenes detail, including every script edit, camera angle, and production choice.

> Trace-level logs can reveal performance bottlenecks and optimization opportunities that might be invisible at higher logging levels:

> Monitoring Resource Utilization: Trace logs can show how resources like memory and CPU are used, helping you optimize them for better performance.

https://newrelic.com/blog/how-to-relic/why-you-need-trace-level-logging

Reply:
Once again, I think there's philosophical differences at play.  I read your evidence and conclude "INFO", you read them and conclude "TRACE".  We're both smart people and sometimes we'll disagree, and that's ok.  There probably is no 100% "correct" answer, like tabs versus spaces, etc.

In the end, I'll set the level to whatever the team decides, and hope that we'll always include the new diagnostic output in the production logs, because a lot of the value of that information is that it's available after an unexpected failure of unknown origin.

Reply:
So far, there's two "votes" for INFO level (me and the "approved" review) and one "vote" for TRACE level, so if we average them and round to nearest, we end up with DEBUG level.  So, I changed the `log` method to user the `Logger.debug` method.  Not complaining, but it's worth noting that it's a relatively costly change, because now, to merge, I'll need to update the logging config in compose_setup to ensure that the information shows up, do the compose_setup/dockstore-deploy PR dance again, and repeat testing on qa...